demo/src/scala/tuiexamples/UserInputExample.scala

```diff
@@ -99,44 +94,42 @@ object UserInputExample {
           Style.DEFAULT
         )
     }
-    val text = msg.overwrittenStyle(style)
-
-    val help_message = ParagraphWidget(text = text)
-    f.renderWidget(help_message, chunks(0))
-
-    val input = ParagraphWidget(
-      text = Text.nostyle(app.input),
-      block = Some(BlockWidget(borders = Borders.ALL, title = Some(Spans.nostyle("Input")))),
-      style = app.input_mode match {
-        case InputMode.Normal  => Style.DEFAULT
-        case InputMode.Editing => Style.DEFAULT.fg(Color.Yellow)
-      }
-    )
-    f.renderWidget(input, chunks(1))
-
-    app.input_mode match {
-      case InputMode.Normal =>
-        // Hide the cursor. `Frame` does this by default, so we don't need to do anything here
-        ()
-
-      case InputMode.Editing =>
-        // Make the cursor visible and ask tui-rs to put it at the specified coordinates after rendering
-        f.setCursor(
-          // Put cursor past the end of the input text
-          x = chunks(1).x + Grapheme(app.input).width + 1,
-          // Move one line down, from the border to the input line
-          y = chunks(1).y + 1
-        )
-    }
-
     val items: Array[ListWidget.Item] =
       app.messages.zipWithIndex.map { case (m, i) => ListWidget.Item(content = Text.nostyle(s"$i: $m")) }
 
-    val messages =
-      ListWidget(
-        items = items,
-        block = Some(BlockWidget(borders = Borders.ALL, title = Some(Spans.nostyle("Messages"))))
+    Layout
+      .detailed(direction = Direction.Vertical, margin = Margin(2))(
+        Constraint.Length(1) -> ParagraphWidget(text = msg.overwrittenStyle(style)),
+        Constraint.Length(3) -> Widget { (area, buf) =>
+          BlockWidget(
+            title = Some(Spans.nostyle("Input")),
+            borders = Borders.ALL,
+            style = app.input_mode match {
+              case InputMode.Normal  => Style.DEFAULT
+              case InputMode.Editing => Style.DEFAULT.fg(Color.Yellow)
+            }
+          )(ParagraphWidget(text = Text.nostyle(app.input)))
+            .render(area, buf)
```

Comment:
Alright so I've been playing around with this a bit, and this is one of the things that I think may really trip people up and doesn't feel super great. Basically I don't think it's always clear to the user when they need to call `render` or not. I think this `Layout` here actually illustrates it well. I think an average user getting used to this would probably do the following.

1. Ok, so I see a `Layout` takes a `(constraint, widget)*`. Sure that makes sense
2. So they see the first constraint as well as the `ParagraphWidget`. Alright, that also tracks
3. Now we have a new `Widget` here acting as a container for the `BlockWidget` which also has children. I think this is where the first part of the confusion enters. A `BlockWidget` is also a widget, so why do I need to wrap it in a `Widget`?
4. This `BlockWidget` also has a `ParagraphWidget` as a child and we call `render` on it? I think here another point of confusion comes in, why do we now have to call render on this widget when we didn't on the widget in the first constraint?
5. Next constraint we now have a block widget that is not wrapped in a `Widget` and we _don't_ call render on it. So again, maybe confusing to the user why now we have another `BlockWidget`, but this time it doesn't need to be in a `Widget` and there is no `.render` necessary here either.
6. Finally we call `render` on the whole thing

All that to say without some knowledge of internal details it might be tricky to a user to know when they need to call render or not, and when they should wrap in a `Widget`.

Maybe just having clear dos about why a `Widget` would help, but I also sort of wish that you could just call `render` once on the `Layout` and have it cascade render everything inside of it. Thoughts?

Reply:
Btw looking closer I see the issue is in `Widget` since the `Layout` calls `render` but it's just `()` so we have to call `render` on whatever widget is inside of the `Widget`. I think if it's possible figure out a way to not have to do that would really help this situation.